Bump openwebifpy to 4.2.0

autinerd opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

Bump openwebifpy to 4.2.0 to fix the issue #107390
https://github.com/autinerd/openwebifpy/compare/4.0.4...4.2.0

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [X] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #107390
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] I have followed the [perfect PR recommendations][perfect-pr]
- [X] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [X] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [X] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [X] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [X] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

autinerd committed:
use dict instead of custom class to hold data from enigma2 API

autinerd committed:
fix #107449

MartinHjelmare commented on homeassistant/components/enigma2/media_player.py:
```diff
@@ -136,7 +143,15 @@ def __init__(self, name: str, device: OpenWebIfDevice, about: dict) -> None:
-        await self._device.turn_off()
+        try:
+            await self._device.turn_off()
+        except ServerDisconnectedError as err:
```
Are these exceptions new in the bumped library? Otherwise please move it to a separate PR.

autinerd replied:
Okay, I have reverted that change and will put it in a separate PR. Now it is only the neccessary changes to fix #107390 and the library bump to do that.

autinerd committed:
Revert changes in last commit

MartinHjelmare commented:
What was the caching and access issue?

autinerd commented:
Currently the data from the API is stored in a custom class `OpenWebifStatus`, accessible by `self._device.status`. For some reason, when multiple devices are updating, there are issues, that the object contains not the right data.

I have 3 devices, `gbquad4k`, `vuduo2` and `osmini`. The first two devices are on, the last is off.

Example log when everything works (with print statements in the `async_update` function and the log statements from the library, only printing the data when off, because when on it is quite a bit larger):

```
2024-01-06 21:26:07.436 INFO (MainThread) [openwebif.api] Got 200 from: http://gbquad4k:80/api/statusinfo
gbquad4k: Changed state to on
2024-01-06 21:26:07.457 INFO (MainThread) [openwebif.api] Got 200 from: http://vuduo2:80/api/statusinfo
vuduo2: Changed state to on
2024-01-06 21:26:07.671 INFO (MainThread) [openwebif.api] Got 200 from: http://osmini:80/api/statusinfo
osmini: Changed state to off, received {'inStandby': 'true', 'currservice_begin': '', 'muted': True, 'isRecording': 'false', 'isStreaming': 'false', 'currservice_description': '', 'volume': 80, 'currservice_fulldescription': 'N/A', 'currservice_name': 'N/A', 'currservice_filename': '', 'transcoding': False, 'currservice_end': '', 'currservice_id': -1, 'Streaming_list': ''}
```

Example log, when the problem happens:

```
2024-01-06 21:27:17.441 INFO (MainThread) [openwebif.api] Got 200 from: http://gbquad4k:80/api/statusinfo
gbquad4k: Changed state to on
2024-01-06 21:27:17.460 INFO (MainThread) [openwebif.api] Got 200 from: http://vuduo2:80/api/statusinfo
2024-01-06 21:27:17.505 INFO (MainThread) [openwebif.api] Got 200 from: http://osmini:80/api/statusinfo
vuduo2: Changed state to off, received {'inStandby': 'true', 'currservice_begin': '', 'muted': True, 'isRecording': 'false', 'isStreaming': 'false', 'currservice_description': '', 'volume': 80, 'currservice_fulldescription': 'N/A', 'currservice_name': 'N/A', 'currservice_filename': '', 'transcoding': False, 'currservice_end': '', 'currservice_id': -1, 'Streaming_list': ''}
osmini: Changed state to off, received {'inStandby': 'true', 'currservice_begin': '', 'muted': True, 'isRecording': 'false', 'isStreaming': 'false', 'currservice_description': '', 'volume': 80, 'currservice_fulldescription': 'N/A', 'currservice_name': 'N/A', 'currservice_filename': '', 'transcoding': False, 'currservice_end': '', 'currservice_id': -1, 'Streaming_list': ''}
```

When the third device receives data before the second device is being updated in `async_setup`, then the last two devices are getting assigned the same data, even though inside the library each device has their own data. (I used the python debugger, and in the library scope `self.status` always contains the correct data for each device, but in the hass scope `self._device.status` contains the data from the other device, even though for example `self._device.mac_address` is always the correct device)

When using a dict for the data (like now with `self._device.status_info`) everything works flawlessly. (I wanted to switch to a dict anyway when implementing the DataUpdateCoordinator, so it is not too bad for me, but normally I would have done that while implementing the DataUpdateCoordinator and not now)

autinerd commented:
The dict returned in the logs are accessed by `self._device.status.status_info`, so an attribute of the `OpenWebifStatus` class

MartinHjelmare commented:
The problem is that you initialize a class attribute to a mutable object:

https://github.com/autinerd/openwebifpy/blob/128123d56730427c2fd4d2aaa9368837f75e320b/openwebif/api.py#L82C6-L82C6

That means that all instances of the device class will hold the same object on the `status` attribute and l instances will see updates to that object.

The correct solution is to initialize the attribute in the init method of the device class.

Generally we prefer dataclass over dict since the former is typed and has a nicer user interface with attributes compared to just key access. I think you should just fix the issue and go back to the dataclass.

MartinHjelmare commented:
You have the same problem currently with the dict:

https://github.com/autinerd/openwebifpy/blob/e472aa85b9e948eed8646e9767715db26ebdfc8b/openwebif/api.py#L56

autinerd commented:
The dict is inside the class, yes, that is the problem. That is why I have now moved the dict outside of the `OpenWebifStatus` class directly to the device class, where it works without problems.

MartinHjelmare commented:
It's only working by chance currently since you replace the attribute value instead of updating the dict when the device updates. If you'd update the dict you'd see the same problem.

It's bad practice to initialize a class attribute to a mutable object unless that feature is explicitly required as it's very error prone.

Please initialize the attribute in the `__init__` method. I also suggest reverting to a dataclass since that's a much nicer interface.

autinerd committed:
Revert "use dict instead of custom class to hold data from enigma2 API"

This reverts commit 9a67916eeeedd22d062cee489b0b786c3ab4ecf0.

autinerd committed:
Bump to 4.2.0

autinerd commented:
Thank you very much, that was indeed the problem, it works now with the dataclass as well.

autinerd committed:
Merge branch 'dev' into enigma2-2024-003
